### PR TITLE
Threads 17: Show parallel response lifecycle states

### PR DESCRIPTION
**Pull request by FranciscoMoretti**
## Summary
- Distinguishes queued, generating, and complete parallel responses.
- Keeps loading UI and the stop control active for optimistic and streaming cards.
- Hides copy, feedback, retry, and follow-up actions until the selected response settles.
- Synchronizes the compatibility store to `submitted` at the app submission boundary.

## Behavior
Parallel cards no longer flash `Task completed`, show premature actions, or expose a send icon while work is pending.

## Verification
- Browser verified first-message and follow-up parallel cards with GPT-5 mini and GPT-5 nano.
- Browser verified selected-run stop isolation.
- 114 app unit tests pass at the stack tip.

## Screenshot
Parallel response cards

## Review focus
The single response-lifecycle rule used by cards, actions, loading, and the composer.

---
## Summary by cubic
Shows accurate lifecycle states for parallel responses and blocks input/actions while any response is active.

- Parallel cards derive queued, generating, complete, stopped, and failed states from the stream marker, run status, and selection; failed states override stale stream markers, and completion actions stay hidden until the stream marker clears.
- Queued card slots are selectable and auto-navigate to the response when streaming starts.
- Stopping settles the selected response immediately: it clears the stream marker for active streams and removes `pending:` responses before the server creates them.
- Loading skeletons stay for queued and pending states until the stream marker clears.

**Migration**
- Replace removed `useChatActions` methods (`setMessages`, `pushMessage`, `popMessage`, `replaceMessage`, `replaceMessageById`, `setStatus`, `setError`, `setId`, `setNewChat`, `reset`) and `useAddMessageToTree` with thread/store helpers; `useLastMessageMetadata` now exposes the last message's metadata for lifecycle-aware UI.

<sup>Written for commit 6f5284d7e423f1a7dea127a461ab3435a66ceed0. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/FranciscoMoretti/chat-js/pull/247?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by Sourcery

Coordinate parallel response lifecycle state across the interface so queued, active, and completed responses behave consistently.

New Features:
- Distinguish queued, generating, and completed states for parallel response cards.
- Keep queued and active responses in loading states while gating composer interactions and completion actions until responses settle.
- Allow queued response slots to be selected and navigate to them when streaming begins.

Bug Fixes:
- Prevent premature completion labels, response actions, send controls, and loading-state transitions for parallel responses.
- Ensure stopping a selected response settles only that response and avoids unnecessary server calls for queued responses.

Enhancements:
- Centralize response-aware lifecycle handling across cards, loading indicators, actions, and the composer.
- Expose last-message metadata for lifecycle-aware UI state and simplify chat action APIs in favor of thread/store helpers.

Tests:
- Add unit coverage for parallel response lifecycle labels, response-aware status, and isolated stop behavior.

### CODE REVIEW

**Comment by FranciscoMoretti** [comment]
* **#254** <a href="https://app.graphite.com/github/pr/FranciscoMoretti/chat-js/254?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#253** <a href="https://app.graphite.com/github/pr/FranciscoMoretti/chat-js/253?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#252** <a href="https://app.graphite.com/github/pr/FranciscoMoretti/chat-js/252?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#251** <a href="https://app.graphite.com/github/pr/FranciscoMoretti/chat-js/251?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#250** <a href="https://app.graphite.com/github/pr/FranciscoMoretti/chat-js/250?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#249** <a href="https://app.graphite.com/github/pr/FranciscoMoretti/chat-js/249?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#248** <a href="https://app.graphite.com/github/pr/FranciscoMoretti/chat-js/248?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#247** <a href="https://app.graphite.com/github/pr/FranciscoMoretti/chat-js/247?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/FranciscoMoretti/chat-js/247?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#246** <a href="https://app.graphite.com/github/pr/FranciscoMoretti/chat-js/246?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#245** <a href="https://app.graphite.com/github/pr/FranciscoMoretti/chat-js/245?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#244** <a href="https://app.graphite.com/github/pr/FranciscoMoretti/chat-js/244?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#243** <a href="https://app.graphite.com/github/pr/FranciscoMoretti/chat-js/243?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#242** <a href="https://app.graphite.com/github/pr/FranciscoMoretti/chat-js/242?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#241** <a href="https://app.graphite.com/github/pr/FranciscoMoretti/chat-js/241?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#240** <a href="https://app.graphite.com/github/pr/FranciscoMoretti/chat-js/240?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#239** <a href="https://app.graphite.com/github/pr/FranciscoMoretti/chat-js/239?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#238** <a href="https://app.graphite.com/github/pr/FranciscoMoretti/chat-js/238?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#237** <a href="https://app.graphite.com/github/pr/FranciscoMoretti/chat-js/237?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#236** <a href="https://app.graphite.com/github/pr/FranciscoMoretti/chat-js/236?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#235** <a href="https://app.graphite.com/github/pr/FranciscoMoretti/chat-js/235?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#234** <a href="https://app.graphite.com/github/pr/FranciscoMoretti/chat-js/234?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: apps/chat/components/parallel-response-status.ts [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+  }
+  return isSelected ? "Selected" : "Task completed";
+}
+
+export function getResponseAwareStatus(
+  status: ChatStatus,
+  message: ParallelResponseStatusMessage | null
+): ChatStatus {
+  const activeStreamId = message?.metadata.activeStreamId;
+  if (!activeStreamId || status === "submitted" || status === "streaming") {
```
[the remark this answers is not in the document]
Comment (FranciscoMoretti) [reply apps/chat/components/parallel-response-status.ts]: Fixed in fe1550cf. Metadata now supplements status only while AI SDK reports ready; submitted, streaming, and error remain authoritative. A regression test covers error with both pending and real stream markers.

[the remark this answers is not in the document]
Comment (FranciscoMoretti) [reply apps/chat/components/parallel-response-status.ts]: Fixed in fe1550cf. getResponseAwareStatus now returns every non-ready AI SDK status unchanged, so a stale stream marker cannot override error.

File: apps/chat/lib/stores/hooks-base.ts [hunk 2]
```diff
@@ -95,3 +95,6 @@ export const useMessageMetadataById = (
 
 export const useLastMessageId = () =>
   useBaseChatStore((state) => state.getLastMessageId());
+
+export const useLastMessageMetadata = () =>
+  useBaseChatStore((state) => state.getThrottledMessages().at(-1)?.metadata);
```
[the remark this answers is not in the document]
Comment (FranciscoMoretti) [reply apps/chat/lib/stores/hooks-base.ts]: Fixed in fe1550cf. useLastMessageMetadata now uses shallow equality, matching the other metadata selector.

File: apps/chat/components/parallel-response-cards.test.ts [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    );
+    assert.equal(
+      getResponseAwareStatus("ready", createAssistantMessage("stream-1")),
+      "streaming"
+    );
+    assert.equal(
+      getResponseAwareStatus("ready", createAssistantMessage(null)),
+      "ready"
+    );
+  });
```
[the remark this answers is not in the document]
Comment (FranciscoMoretti) [reply apps/chat/components/parallel-response-cards.test.ts:71]: Added in fe1550cf. The test covers error with both pending: and real activeStreamId values and verifies error is preserved.

File: apps/chat/components/parallel-response-status.ts [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    return status;
+  }
+
+  return activeStreamId.startsWith("pending:") ? "submitted" : "streaming";
```
[the remark this answers is not in the document]
Comment (FranciscoMoretti) [reply apps/chat/components/parallel-response-status.ts:63]: Fixed in fe1550cf. Only ready may be promoted from a stream marker; error now remains error and the composer exposes its retry/send path.

File: apps/chat/components/parallel-response-cards.tsx [hunk 5]
```diff
@@ -117,7 +118,23 @@ function PureParallelResponseCards({ messageId }: { messageId: string }) {
...
+    const response = cardSlots.find(
+      (slot) => slot.parallelIndex === pendingParallelIndex
+    )?.message;
+    if (!response) {
+      return;
+    }
+
+    setPendingParallelIndex(null);
+    navigateToMessage(response.id);
+  }, [cardSlots, navigateToMessage, pendingParallelIndex]);
```
[the remark this answers is not in the document]
Comment (FranciscoMoretti) [reply apps/chat/components/parallel-response-cards.tsx]: Addressed by the current run-aware lifecycle implementation. A message-less run derives Failed or Stopped from its terminal run status, so it no longer renders a spinner or Generating. The selected index intentionally remains latched because setActiveRun selected that exact terminal run; clearing it would make the card highlight disagree with chat.status/chat.error and stop targeting.

File: apps/chat/components/multimodal-input.tsx [hunk 6]
```diff
@@ -615,17 +628,34 @@ function PureMultimodalInput({
...
+    if (session?.user && lastMessageId && !isPendingResponse) {
       stopStreamMutation.mutate({ chatId, messageId: lastMessageId });
     }
     stopHelper?.();
-  }, [chatId, lastMessageId, session?.user, stopHelper, stopStreamMutation]);
+    if (lastMessageId) {
+      setMessages(
+        clearResponseActiveStream(storeApi.getState().messages, lastMessageId)
+      );
+    }
```
[the remark this answers is not in the document]
Comment (FranciscoMoretti) [reply apps/chat/components/multimodal-input.tsx:636]: Fixed in 6f5284d7 with a stronger transition than the suggested guard. A stopped pending response has no emitted server response, so the optimistic leaf is removed from the tree. Emitted responses keep their partial content and clear only a real stream marker. clearResponseActiveStream also refuses pending: markers, with a regression test, so an empty optimistic response cannot be mislabeled complete or expose completion actions.

**Comment by FranciscoMoretti** [comment]
@greptileai

**Comment by FranciscoMoretti** [comment]
@greptile review

**Comment by FranciscoMoretti** [comment]
@greptile-apps review

**Comment by FranciscoMoretti** [comment]
@greptile review
